File: LEETCODE_2316_Count_Unreachable_Pairs_of_Nodes_in_an_Undirected_Graph.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
    int bfs(int& source, vector<bool>& visited, unordered_map<int,vector<int>>& graph) {
        int count = 0;
        queue<int> q;
        q.push(source);
        visited[source] = true;
        while(!q.empty()) {
            int node = q.front();
            q.pop();
            count++;
            for(int& neighbour: graph[node]) {
                if(visited[neighbour] == false) {
                    q.push(neighbour);
                    visited[neighbour] = true;
                }
            }
        }
        return count;
    }
public:
    long long countPairs(int n, vector<vector<int>>& edges) {
        unordered_map<int,vector<int>> graph;
        for(vector<int>& edge: edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }
        int component_sizes = 0;
        vector<bool> visited(n,false);
        unsigned long long int answer = 0;
        for(int node = 0; node < n; node++) {
            if(visited[node] == false) {
                int size = this->bfs(node, visited, graph);
                answer += (component_sizes*1LL * size*1LL);
                component_sizes += size;
            }
        }
        return answer;
    }
};

File: LEETCODE_2316_Count_Unreachable_Pairs_of_Nodes_in_an_Undirected_Graph.java
import java.util.*;

class LEETCODE_2316_Count_Unreachable_Pairs_of_Nodes_in_an_Undirected_Graph {

    private int bfs(int source, boolean[] visited, Map<Integer, List<Integer>> graph) {
        ArrayDeque<Integer> q = new ArrayDeque<>();
        q.offerLast(source);
        visited[source] = true;
        int count = 0;
        while(!q.isEmpty()) {
            int node = q.pollFirst();
            count++;
            List<Integer> neighbours = new ArrayList<>();
            if(graph.containsKey(node)) {
                neighbours = graph.get(node);
            }
            for(int neighbour: neighbours) {   // "java.util.Map.get(Object)" is null
                if(!visited[neighbour]) {
                    q.offerLast(neighbour);
                    visited[neighbour] = true;
                }
            }
        }
        return count;
    }

    public long countPairs(int n, int[][] edges) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for(int[] edge: edges) {
            if(!graph.containsKey(edge[0])) {
                graph.put(edge[0], new ArrayList<>());
            }
            if(!graph.containsKey(edge[1])) {
                graph.put(edge[1], new ArrayList<>());
            }
            graph.get(edge[0]).add(edge[1]);
            graph.get(edge[1]).add(edge[0]);
        }
        boolean[] visited = new boolean[n];
        Arrays.fill(visited, false);
        long answer = 0, component_sizes = 0;
        for(int node = 0; node < n; node++) {
            if(!visited[node]) {
                long size = this.bfs(node, visited, graph);
                answer += component_sizes * size;
                component_sizes += size;
            }
        }
        return answer;
    }
}